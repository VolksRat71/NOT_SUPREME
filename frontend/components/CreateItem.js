import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import Router from 'next/router';

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION (
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ){
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
    `

class CreateItem extends Component {
    state = {
        title: 'Mask Cat',
        description: 'Judgy Bitch will start at you.',
        image: 'https://image.shutterstock.com/image-photo/cat-medical-mask-protective-antiviral-260nw-1684423789.jpg',
        largeImage: 'https://image.shutterstock.com/image-photo/cat-medical-mask-protective-antiviral-260nw-1684423789.jpg',
        price: 20
    }

    handleChange = e => {
        const { name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value
        this.setState({ [name]: val })
    }

    uploadFile = async e => {
        console.log("uploading file ...");
        const files = e.target.files;
        const data = new FormData();
        data.append('file',  files[0]);
        data.append('upload_present', 'NotSupreme');

        const res = await fetch(
            'https://api.cloudinary.com/v1_1/notsupreme/image/upload',
            {
                method: 'POST',
                body: data
            }
        )

        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url
        });
    }

    render() {
        return (
            <Mutation
                mutation={CREATE_ITEM_MUTATION}
                variables={this.state}>
                {(createItem, { loading, error }) => (
                    <Form onSubmit={async e => {
                        e.preventDefault();
                        const res = await createItem();
                        Router.push({
                            pathname: '/item',
                            query: { id: res.data.createItem.id }
                        })
                    }}>
                        <Error error={error} />
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor="file">
                                Image
                                <input
                                    type="file"
                                    id="file"
                                    name="file"
                                    placeholder="Upload an image"
                                    required
                                    onChange={this.uploadFile}
                                />
                            </label>
                            <label htmlFor="title">
                                Title
                                <input
                                    type="text"
                                    id="title"
                                    name="title"
                                    placeholder="Title"
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="price">
                                Price
                                <input
                                    type="number"
                                    id="price"
                                    name="price"
                                    placeholder="Price"
                                    required
                                    value={this.state.price}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="description">
                                Description
                                <textarea
                                    id="description"
                                    name="description"
                                    placeholder="Enter A Description"
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <button
                                type="submit"
                            >Submit</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        )
    }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };