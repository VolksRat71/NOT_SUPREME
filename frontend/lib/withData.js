import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    // Local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cart open value from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);