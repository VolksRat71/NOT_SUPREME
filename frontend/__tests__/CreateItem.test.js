import { mount } from 'enzyme';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { fakeItem } from '../lib/testUtils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';

const dogImage = 'https://dog.com/dog.png'

// * mock the global fetch res from cloudinary API
global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: dogImage,
        eager: [{ secure_url: dogImage }]
    })
})

describe('<CreateIem/> ', () => {
    test('Renders and matches snapshot ', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        )
        const form = wrapper.find('form[data-test="form"]')
        expect(toJSON(form)).toMatchSnapshot()
    })

    test('Form uploads file when changed ', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        )
        const input = wrapper.find('input[type="file"]')
        input.simulate('change', {
            target: { files: { files: ['fakePicOfDog.png'] } }
        });
        await wait()
        const component = wrapper.find('CreateItem').instance();
        expect(component.state.image).toBe(dogImage);
        expect(component.state.largeImage).toBe(dogImage);
        expect(global.fetch).toHaveBeenCalled();
        //! RESET THE GLOBAL FETCH WHEN TEST IS COMPLETE!!
        global.fetch.mockReset()
    })

    test('Handle state change ', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem />
            </MockedProvider>
        )
        wrapper
            .find('#title')
            .simulate('change', {
                target: { value: 'Testing', name: 'title' }
            })
        wrapper
            .find('#price')
            .simulate('change', {
                target: { value: 50000, name: 'price', type: 'number' }
            })
        wrapper
            .find('#description')
            .simulate('change', {
                target: { value: 'This is an Item', name: 'description' }
            })

        const state = wrapper.find('CreateItem').instance().state;
        expect(state).toMatchObject({
            title: 'Testing',
            description: 'This is an Item',
            image: '',
            largeImage: '',
            price: 50000
        })
    })

    xtest('Creates and Item when the form is submitted ', async () => {
        const { title, description, price } = fakeItem();
        const mocks = [{
            request: {
                query: CREATE_ITEM_MUTATION,
                variables: {
                    title: title,
                    description: description,
                    image: '',
                    largeImage: '',
                    price: price
                }
            },
            result: {
                data: {
                    createItem: {
                        ...fakeItem(),
                        id: 'abc123',
                        __typename: 'Item'
                    }
                }
            }
        }]
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateItem />
            </MockedProvider>
        );
        wrapper
            .find('#title')
            .simulate('change', {
                target: { value: 'Testing', name: 'title' }
            })
        wrapper
            .find('#price')
            .simulate('change', {
                target: { value: 50000, name: 'price', type: 'number' }
            })
        wrapper
            .find('#description')
            .simulate('change', {
                target: { value: 'This is an Item', name: 'description' }
            })
        Router.router = { push: jest.fn() };
        wrapper.find('form').simulate('submit');
        await wait(50);
        // TODO: Again, Why are we throwing an error on component
        expect(Router.router.push).toHaveBeenCalled()
    })
})

