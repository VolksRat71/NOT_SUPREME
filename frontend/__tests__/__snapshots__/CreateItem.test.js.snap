// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateIem/>  Creates and Item when the form is submitted  1`] = `
<fieldset
  aria-busy={false}
  disabled={false}
>
  <label
    htmlFor="file"
  >
    Image
    <input
      id="file"
      name="file"
      onChange={[Function]}
      placeholder="Upload an image"
      required={true}
      type="file"
    />
  </label>
  <label
    htmlFor="title"
  >
    Title
    <input
      id="title"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value="Testing"
    />
  </label>
  <label
    htmlFor="price"
  >
    Price
    <input
      id="price"
      name="price"
      onChange={[Function]}
      placeholder="Price"
      required={true}
      type="number"
      value={50000}
    />
  </label>
  <label
    htmlFor="description"
  >
    Description
    <textarea
      id="description"
      name="description"
      onChange={[Function]}
      placeholder="Enter A Description"
      required={true}
      value="This is an Item"
    />
  </label>
  <button
    type="submit"
  >
    Submit
  </button>
</fieldset>
`;

exports[`<CreateIem/>  Renders and matches snapshot  1`] = `
<form
  className="Form-sc-17mrryo-0 kCLdEE"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter A Description"
        required={true}
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
